### ISSUE TITLE
invoking `:init.stop` takes 5 seconds if at least one websocket client is connected

### ISSUE THREAD

**Issue by sasa1977** [phoenixframework/phoenix#2466]
### Environment

* Elixir version (elixir -v): 1.5.0
* Phoenix version (mix deps): 1.3.0

### Expected behavior

Invoking `:init.stop` should properly terminate idle socket process as soon as possible.

### Actual behavior

If at least one client is connected over websocket, invoking `:init.stop` takes 5 seconds until it kills the socket process brutally:

```
mix phx.new test --no-ecto
cd test
iex -S mix phx.server

# open a browser and navigate to localhost:4000

iex> :init.stop
# the system stops after 5 seconds
```

If no websocket client is connected, the system stops almost immediately.

I didn't test the behaviour with long polling.

### More details

I believe the issue originates [here](https://github.com/phoenixframework/phoenix/blob/b98e8c8a33adb7521118c9bce2d6b93989075b5a/lib/phoenix/transports/websocket.ex#L108). The socket process is trapping exits, so by default it won't terminate when it receives a shutdown signal from its parent.

When handling the `:EXIT` message, [the socket process only reacts to an exit of a child](https://github.com/phoenixframework/phoenix/blob/b98e8c8a33adb7521118c9bce2d6b93989075b5a/lib/phoenix/transports/websocket.ex#L136-L137). Hence the exit sent from the parent is ignored, and the process does nothing. It will ultimately be terminated after the timeout happens in the parent, but we're waiting for 5 seconds for no real purpose.

At surface, the solution would be to detect if the `:EXIT` process originated from the parent, and if so, issue the same exit signal to children (channel processes), wait for them to terminate, or kill them brutally if they refuse to do so (e.g. they are trapping exits).

One problem is that it seems there's no information in the socket process about the parent. I believe this is due to how cowboy (or more precisely ranch) starts socket processes. This could be solved with some trickery, e.g. if the cowboy handler used `Process.info(self(), [:links])` during init phase to determine the parent (the only process which is linked). It's a hack, but it should do the job.

One thing which would be really nice is to allow channels to send some final info before they terminate. I think this might turn out to be quite tricky so maybe that should be tackled separately.

**Comment by chrismccord** [comment]
Related to https://github.com/phoenixframework/phoenix/issues/2432 , in that if we properly start and supervise  channel processes, then this issue is mostly fixed, I think.

**Comment by sasa1977** [comment]
If in the process of moving the channel processes under a dedicated supervisor you remove exit trapping from the socket process, then I believe the immediate issue would be fixed. 

However, if the socket process is not trapping exits, then I think it will be possible (though not probable) that a socket process is taken down, while the channels are left running. This could happen if the parent of the socket process decides to stop. The parent would send an exit signal to the socket process, which would immediately stop (it's not trapping exits anymore). The exit signal would be propagated to each channel process. Now if the channel processes are trapping exits (or use monitors to detect socket process termination), then they wouldn't be stopped immediately, and no other process would enforce their termination. Under some conditions they might run for quite a long time (maybe forever).

If the socket process traps exits, then the issue won't be fixed.

**Comment by josevalim** [comment]
I would say we should solve and test this decoupled from the other decision
because of the reasons given by Sasa.
-- 


*José Valim*
www.plataformatec.com.br
Skype: jv.ptec
Founder and Director of R&D

**Comment by scrogson** [comment]
I'm going to start working on a fix for this.

**Comment by josevalim** [comment]
Fixed by #2861.
